File: cpp/graphs/dfs.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs(int start, vector<int> &visited, vector<int> adj[], vector<int> &ans) {
  visited[start] = 1;
  ans.push_back(start);
  for (auto n : adj[start]) {
    if (!visited[n]) {
      dfs(n, visited, adj, ans);
    }
  }
}

vector<int> dfsGraph(int start, int n, vector<int> adj[]) {
  vector<int> visited(n + 1, 0);
  vector<int> ans;

  dfs(start, visited, adj, ans);
  return ans;
}

int main() {

  int n, m;
  cout << "Enter no. of vertices: ";
  cin >> n;
  cout << "Enter no. of edges: ";
  cin >> m;

  vector<int> adj[n + 1]; // assuming the starting node as 1

  while (m--) {
    int u, v;
    cin >> u >> v;
    adj[u].push_back(v);
    adj[v].push_back(u); // not needed for directed graph (u -> v)
  }
  // incase of weighted graph store pairs with weights
  // (node, weight), (node1, weight)..
  //
  int start = 1;
  vector<int> res = dfsGraph(start, n, adj);
  for (auto ele : res) {
    cout << ele << " ";
  }
  cout << endl;

  return 0;
}

File: python/graphs/dfs.py
from listRepresentation import generate_adj_list


def dfs_traverse(start, n, adj_list):
    visited = [0] * (n + 1)
    res = []

    def dfs(node, adj_list):
        nonlocal visited, res
        visited[node] = 1
        res.append(node)

        for v in adj_list[node]:
            if not visited[v]:
                dfs(v, adj_list)

    dfs(start, adj_list)
    return res

# edges = int(input("Enter no. of edges: "))
# vertices = int(input("Enter no. of vertices: "))
#
#
# nodes = []
# for _ in range(edges):
#     values = input("").split()
#     u = int(values[0])
#     v = int(values[1])
#     nodes.append((u, v))
#


nodes_data = [(1, 2), (1, 3), (2, 4), (2, 5), (3, 6), (3, 7), (7, 8)]
edges = 7
nodes = 8
adj_list = generate_adj_list(nodes_data)

print(dfs_traverse(1, nodes, adj_list))
